### ISSUE TITLE
Vm cannot run programs with % in name

### ISSUE THREAD

**Issue by sigurdm** [dart-lang/sdk#41914]
To reproduce reate `bin/ab%22.dart` with a small program, and run `dart bin/ab%22.dart`

prints:
```
Error: Error when reading 'bin/ab%22.dart': No such file or directory
```

The same happens for snapshots.

My motivation is that I would like to url-encode the sdk-version into the name of a snapshot(for pub run) to avoid running a bad snapshot.

**Comment by natebosch** [comment]
> My motivation is that I would like to url-encode the sdk-version into the name of a snapshot(for pub run) to avoid running a bad snapshot.

I feel strongly that we should not be trying to hack around a bug in the SDK, pub is not the only place we use this pattern and we should not go rewrite them all to encode the SDK version in a file name. Have we filed an issue for not getting the correct exception on a bad snapshot?

**Comment by jakemac53** [comment]
This still seems like a valid bug though right?

**Comment by natebosch** [comment]
> This still seems like a valid bug though right?

Yes.

**Comment by sigurdm** [comment]
> > My motivation is that I would like to url-encode the sdk-version into the name of a snapshot(for pub run) to avoid running a bad snapshot.
> 
> I feel strongly that we should not be trying to hack around a bug in the SDK, pub is not the only place we use this pattern and we should not go rewrite them all to encode the SDK version in a file name. Have we filed an issue for not getting the correct exception on a bad snapshot?

We should not work around a bug - true. But I find that encoding the sdk version in the filename actually it is nicer for `pub run` for the following reasons:

* We are moving to using a subprocess instead of an isolate for pub run (https://github.com/dart-lang/pub/issues/2492). Meaning that there is no exception to catch, we would have to interpret stdout and exitcode, which we can never ensure are not overlapping with what the program itself would return. Alternatively the VM would have a mode that **only** detected validity of the snapshot. But then that would require one more subprocess call before starting the program.
* We can have one snapshot per installed SDK without conflict - maybe not a common use-case, but when you are testing a program with different sdks it works nicely.

For other use-cases we **do** want the VM to throw a known exception on `Isolate.spawnUri`, and it would also be nice to have a way to test a snapshot's validity before running it.
